searchWorker.js

```diff
@@ -1,27 +1,71 @@
-req.responseType = 'text';
-req.send();
-
-const declNames = req.responseText.split('\n');
 
-// Adapted from the default tokenizer and
-// https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5Cp%7BZ%7D&abb=on&c=on&esc=on&g=&i=
-const SEPARATOR = /[._\n\r \u00A0\u1680\u2000-\u200A\u2028\u2029\u202F\u205F\u3000]+/u
+req.open('GET', 'searchable_data.json', false /* blocking */);
+req.responseType = 'json';
+req.send();
 
 importScripts('https://cdn.jsdelivr.net/npm/minisearch@2.4.1/dist/umd/index.min.js');
+    // Include options as per API specs: https://lucaong.github.io/minisearch/modules/_minisearch_.html
 const miniSearch = new MiniSearch({
-    fields: ['decl'],
-    storeFields: ['decl'],
-    tokenize: text => text.split(SEPARATOR),
+    idField: 'name',
+    fields: ['module', 'name', 'description'],
+    storeFields: ['module', 'name', 'description', 'attributes', 'kind']
 });
-miniSearch.addAll(declNames.map((decl, id) => ({decl, id})));
+
+const indexedData = req.response;
+miniSearch.addAll(indexedData);
 
 onconnect = ({ports: [port]}) =>
-    port.onmessage = ({data}) => {
-        const results = miniSearch.search(data.q, {
-            prefix: (term) => term.length > 3,
-            fuzzy: (term) => term.length > 3 && 0.2,
+port.onmessage = ({ data }) => {
+    const { query, filters, maxCount } = data;
+    const filterFunc = (result) => filterItemResult(result, filters);
+
+    if (query && typeof query === "string" && query.length > 0) {
+        const results = miniSearch.search(query, {
+            boost: { module: 1, description: 2, name: 3 },
+            combineWith: 'AND',
+            filter: filterFunc,
+            // prefix: (term) => term.length > 3,
+            // fuzzy: (term) => term.length > 3 && 0.2,
         });
-        port.postMessage(results.slice(0, 20));
-    };
+        
+        const response = typeof maxCount === "number" && maxCount >= 0 ? results.slice(0, maxCount) : results;
+        console.log(response)
+        port.postMessage({response, total: results.length});
+    }
+    port.postMessage({response: [], total: 0});
+};
+
+const filterItemResult = (result, filters = {}) => {
+    const { attributes: attrFilter, kind: kindFilter } = filters;
+    const hasAttrFilter = attrFilter && attrFilter.length > 0;
+    const hasKindFilter = kindFilter && kindFilter.length > 0;
+
+    if (!hasAttrFilter && !hasKindFilter) {
+        return true;
+    }
+    
+    const { attributes: attrRes, kind: kindRes } = result;
+    let isResultAttrIncluded = false;
+    let isResultKindIncluded = false;
+
+    if (hasKindFilter) {
+        isResultKindIncluded = kindFilter.includes(kindRes);
+    }
+
+    if (hasAttrFilter) {
+        for (let attribute of attrRes) {
+            if (attrFilter.includes(attribute)) {
+                isResultAttrIncluded = true;
+                break;
+            }
+        }
+    } 
+
+    return hasKindFilter && hasAttrFilter ? 
```

Comment:
Wouldn't it be better to split these 3 ternary expressions into 3-ifs or early return?

Reply:
The ternary operator by itself is faster than if/else if no additional computation is involved during its execution, but what's more important I think is that this is more readable and easier to change or extend to more types of filters in the long run. I can return the result before getting here, of course - where the check for hasKindFilter is, for example, just return if there is no hasAttrFilter. Also return directly when hasAttrFilter gives a value to isAttrFilterIncluded, if there is no hasKindFilter. But this kind of structuring mixes two different actions: checking if the result matches any of the filters, and deciding whether this is sufficient. To be more exact, if we decide to change the logic from "you need to cover both the attribute filters and the kind filters, if both are present" to "you can cover either filter selected, as long as at least one of any of them matches your attributes or kind" we will have to change the return statement here only. Otherwise we will have to follow through the whole logic in detail and decide whether to move each return statement and how are they connected at all for the purpose of the method itself. I'd leave it like it is, to be honest